Lib/functools.py

```diff
@@ -192,6 +192,222 @@ def total_ordering(cls):
             setattr(cls, opname, opfunc)
     return cls
 
+################################################################################
+### topological sort
+################################################################################
+
+_NODE_OUT = -1
+_NODE_DONE = -2
+
+class _NodeInfo:
+    __slots__ = 'node', 'npredecessors', 'successors'
+
+    def __init__(self, node):
+        # The node this class is augmenting.
+        self.node = node
+
+        # Number of predecessors, generally >= 0. When this value falls to 0,
+        # and is returned by get_ready(), this is set to _NODE_OUT and when the
+        # node is marked done by a call to done(), set to _NODE_DONE.
+        self.npredecessors = 0
+
+        # List of successor nodes. The list can contain duplicated elements as
+        # long as they're all reflected in the successor's npredecessors attribute).
+        self.successors = []
+
+class CycleError(ValueError):
+    pass
+
+class TopologicalSorter:
+    """Provides functionality to topologically sort a graph of hashable nodes"""
+
+    def __init__(self, graph=None):
+        self.node2info = {}
```

Comment:
shouldn't all these be private (aka starting with a ``_``)?

Reply:
Done in aa56145